File: java/Day4.java
import java.io.*;
import java.util.*;

public class Day4 {
    static HashSet<Integer>[] winning;
    static int[][] having;
    static int nCards;
    
    public static void main(String[] args) throws FileNotFoundException {
        long startTime = System.currentTimeMillis();
        Scanner in = new Scanner(new FileReader("input/day4.txt"));
        List<String> input = new ArrayList<>();
        while (in.hasNextLine()) {
            input.add(in.nextLine());
        }
        getCards(input);
        System.out.println("Part I  21821  : " + part1());
        System.out.println("Part II 5539496: " + part2());
        System.out.println("Execution time: " + (System.currentTimeMillis() - startTime) + "ms");
    }
    
    private static int part1() {
        int res = 0;
        for (int i = 0; i < nCards; i++) {
            int ct = 0;
            for (int card : having[i]) {
                if (winning[i].contains(card)) {
                    ct *= 2;
                    if (ct == 0) ct = 1;
                }
            }
            res += ct;
        }
        return res;
    }
    
    private static int part2() {
        Map<Integer, Integer> matches = new HashMap<>();
        for (int i = 0; i < nCards; i++) {
            for (int card : having[i]) {
                if (winning[i].contains(card)) {
                    matches.put(i + 1, matches.getOrDefault(i + 1, 0) + 1);
                }
            }
        }
        
        int res = nCards;
        Queue<Integer> q = new LinkedList<>();
        for (int key : matches.keySet()) {
            q.offer(key);
        }
        while (!q.isEmpty()) {
            int curr = q.poll();
            for (int i = 1; i <= matches.get(curr); i++) {
                if (matches.containsKey(curr + i)) {
                    q.offer(curr + i);
                }
                res++;
            }
        }
        return res;
    }
    
    private static void getCards(List<String> input) {
        nCards = input.size();
        winning = new HashSet[nCards];
        having = new int[nCards][];
        for (int card = 0; card < nCards; card++) {
            String line = input.get(card);
            String[] winAndHave = line.split(": ")[1].split(" \\| ");
            winning[card] = new HashSet<>();
            for (String s : winAndHave[0].split(" ")) {
                if (s.equals("")) continue;
                winning[card].add(Integer.parseInt(s));
            }
            List<Integer> currList = new ArrayList<>();
            for (String s : winAndHave[1].split(" ")) {
                if (s.equals("")) continue;
                currList.add(Integer.parseInt(s));
            }
            having[card] = currList.stream().mapToInt(i -> i).toArray();
        }
    }
}

File: rust/src/bin/day4.rs
use std::collections::{HashSet, HashMap, VecDeque};
use lazy_static::lazy_static;

lazy_static!(
    static ref FILE: String = std::fs::read_to_string("../input/day4.txt").unwrap();
    );

fn main() {
    let (decks, winning_cards) = parse_input(&FILE);
    println!("{}{}", "Expect 21821, Result: ", part1(&decks, &winning_cards));
    println!("{}{}", "Expect 5539496, Result: ", part2(&decks, &winning_cards));
}

fn parse_input(input: &str) -> (Vec<Vec<i32>>, Vec<HashSet<i32>>) {
    let mut winning_cards: Vec<HashSet<i32>> = Vec::new();
    let mut decks: Vec<Vec<i32>> = Vec::new();

    for line in input.lines() {
        let all_cards: Vec<_> = line.splitn(2, ": ").nth(1).unwrap().split(" | ").collect();
        winning_cards.push(all_cards[0].split_whitespace().map(|card| card.parse::<i32>().unwrap()).collect());
        decks.push(all_cards[1].split_whitespace().map(|card| card.parse::<i32>().unwrap()).collect());
    }
    (decks, winning_cards)
}

fn part1(decks: &[Vec<i32>], winning_cards: &[HashSet<i32>]) -> i32 {
    let mut res = 0;
    for (i, deck) in decks.iter().enumerate() {
        let mut ct = 0;
        for card in deck {
            if winning_cards[i].contains(card) {
                ct = if ct == 0 { 1 } else { ct * 2 };
            }
        }
        res += ct;
    }
    res
}

fn part2(decks: &[Vec<i32>], winning_cards: &[HashSet<i32>]) -> i32 {
    let mut matches = HashMap::new();
    for (i, deck) in decks.iter().enumerate() {
        for card in deck {
            if winning_cards[i].contains(card) {
                *matches.entry(i + 1).or_insert(0) += 1;
            }
        }
    }

    let mut res = decks.len() as i32;
    let mut q: VecDeque<_> = matches.keys().cloned().collect();
    while let Some(curr) = q.pop_front() {
        for i in 1..=*matches.get(&curr).unwrap() {
            if matches.contains_key(&(curr + i)) {
                q.push_back(curr + i);
            }
            res += 1;
        }
    }
    res
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn day4_part1() {
        let (decks, winning_cards) = parse_input(&FILE);
        assert_eq!(21821, part1(&decks, &winning_cards))
    }

    #[test]
    fn day4_part2() {
        let (decks, winning_cards) = parse_input(&FILE);
        assert_eq!(5539496, part2(&decks, &winning_cards))
    }
}
